### PR TITLE
Entities should be deleted by the gamemap on client side instead of the render manager

### LINKED ISSUE #247: Entities should be deleted by the gamemap on client side (like on the server side)

### PR DESCRIPTION
**Pull request by hwoarangmy**
- Entities should be deleted by the gamemap on client side instead of the render manager
- Creature reading/writing is the same as rooms/traps
- Small fix for creature sound
- Fixed a memory leak on Creature weapons
- Added a log for entity creation on client side as well as on server side
- Added [Creature][/Creature] in creature list in levels
  fixes #247

### CODE REVIEW

File: source/GameEntity.cpp [hunk 1]
```diff
@@ -57,12 +57,10 @@ void GameEntity::deleteYourself()
...
 
     mIsDeleteRequested = true;
+
     deleteYourselfLocal();
 
-    if(getGameMap()->isServerGameMap())
-    {
-        getGameMap()->queueEntityForDeletion(this);
-    }
+    getGameMap()->queueEntityForDeletion(this);
```
Comment (Bertram25) [inline source/GameEntity.cpp:63]: ahh now I see.

File: source/Creature.cpp [hunk 2]
```diff
@@ -258,209 +265,196 @@ std::string Creature::getFormat()
...
-    is >> tempString;
+    OD_ASSERT_TRUE(is >> tempString);
+    mDefinition = getGameMap()->getClassDescription(tempString);
+    OD_ASSERT_TRUE_MSG(mDefinition != nullptr, "Definition=" + tempString);
+
+    mMaxHP = mDefinition->getMinHp();
+    setHP(mMaxHP);
+    mDigRate = mDefinition->getDigRate();
+    mClaimRate = mDefinition->getClaimRate();
+    mSound = SoundEffectsManager::getSingleton().getCreatureClassSounds(tempString);
```
Comment (Bertram25) [inline source/Creature.cpp:311]: Is that the fix for sound?

File: source/Creature.cpp [hunk 3]
```diff
@@ -107,14 +107,10 @@ Creature::Creature(GameMap* gameMap, CreatureDefinition* definition, bool forceN
     mStatsWindow             (NULL),
     mAttackedTile            (NULL),
     mAttackedObject          (NULL),
-    mSound                   (SoundEffectsManager::getSingleton().getCreatureClassSounds(getName())),
+    mSound                   (SoundEffectsManager::getSingleton().getCreatureClassSounds(definition->getClassName())),
```
Comment (Bertram25) [inline source/Creature.cpp:110]: Ah yes, the definition instead of the name. Nice catch!

Comment (hwoarangmy) [inline source/Creature.cpp:110]: yes, name was not the best here. Moreover, even if it was the name, it is not set in the constructor...

**Comment by Bertram25** [comment]
I've got an assert failure when quitting a map:
EDIT: Nope, I get it also while in game.

```
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,62],Fullness=100
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,62],Fullness=100
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,62],Fullness=100
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,61],Fullness=100
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,61],Fullness=100
ERROR: Assert failed file opendungeons/source/Creature.cpp line 2432 info : 1 creature=Kobold96,tile=[49,61],Fullness=100
```

Seems harmless, though.

The creature is a random kobold, and this is triggered on a random tile position because of the fullness not being 0.

**Comment by Bertram25** [comment]
I don't know whether it is your PR introducing it, though.

**Comment by Bertram25** [comment]
The assert failure apart, this works fine :)

**Comment by hwoarangmy** [comment]
That's not the first time I see this assert. It means that on client side, i creature (in your case, a kobold), is going through a wall. it happens when synchronisation with server is not good enough (we receive the server command too early and we start going to the new position before arriving where we were expected. Usually, the creature is not far from where it should be. But sometimes, it is enough to make it go through a wall.

If you try the empty map, force a creature to go to the top right and display the stat windows, you will see synchronisation is not perfect with server (on server side, you will see the creature arrives before on client - not huge but perceptible). That's why I tried to slow down the server. But if it is too slow, the creatures will sometimes stay idle on client side (waiting for the server) and that would be even worst...

To summarize, if you think this PR is ok, you can merge ^^

**Comment by Bertram25** [comment]
Ok, thanks for the thorough explanation :)

I think I can merge. Since you're the boss in that area, maybe we might have to resync creatures when they are not in sync with the server since too long, or is it already done somehow?

**Comment by hwoarangmy** [comment]
> I think I can merge. Since you're the boss in that area, maybe we might have to resync creatures when they are not in sync with the server since too long, or is it already done somehow?

Not really but IMHO, we should wait to see if there is really a need to resync all as it will produce glitches and might be complex, I think we should not go there until there is a real need.

**Comment by Bertram25** [comment]
ack

**Comment by hwoarangmy** [comment]
@Bertram25 FYI, I've found the reason behind the ASSERT failure:
https://github.com/OpenDungeons/OpenDungeons/blob/development/source/gamemap/GameMap.cpp#L1525

While computing the path through diggable tiles, we try to find the best way considering the creature (Kobold in this case) move speed. But on a diggable tile, fullness can be > 0, which causes the assert fail. I guess it wasn't seen before because we used the creature definition. I will change this in my next PR so that we use ground speed if the tile fullness > 0.

**Comment by Bertram25** [comment]
Great! :D
